### DISCUSSION TITLE
Board driver using stepper_t variables?

### DISCUSSION THREAD

**Discussion by mfp20** [grblHAL/core#52] [Q&A]
I noticed HAL drivers use some of the vars in struct stepper_t. Why?

I mean, from my perspective the HAL should use step_outbits and dir_outbits only, as those are the bits needed to set/clear the gpios. Being an HAL driver means it has to provide simple methods for HW access only.

So, it looks like the stepper_t structure isn't really fit for the role; a left over from grbl conversion to HAL'lized grbl? Any chance to change it?

**Comment by terjeio** [comment]
> I noticed HAL drivers use some of the vars in struct stepper_t. Why?

For making [spindle synced motion](https://github.com/grblHAL/iMXRT1062/blob/a1bf405261fdda240f3adbaaa48ea8e33c1089d0/grblHAL_Teensy4/src/driver.c#L880-L967) work. And some plugins references fields in stepper_t as well.

> So, it looks like the stepper_t structure isn't really fit for the role;

Perhaps not but currently it is, what would an alternative look like?

> a left over from grbl conversion to HAL'lized grbl?

Yes, and I have added fields to it and exposed it (making it accessible via the hal.stepper.pulse_start parameter).

> Any chance to change it?

Yes, but you will have to move code from the HAL to the core to make spindle synced motion work? And new entries has to be added to the HAL struct for the plugins now hooking into hal.stepper.pulse_start? It will be easier to run the relevant HAL/plugin code on the same core (as stepper.c)?

**Reply by mfp20** [reply]
> > I noticed HAL drivers use some of the vars in struct stepper_t. Why?
> 
> For making [spindle synced motion](https://github.com/grblHAL/iMXRT1062/blob/a1bf405261fdda240f3adbaaa48ea8e33c1089d0/grblHAL_Teensy4/src/driver.c#L880-L967) work. And some plugins references fields in stepper_t as well.

Well, plugins are fine as, once installed, they become part of the core logic. The HAL only is ... not completely abstracted.

> > So, it looks like the stepper_t structure isn't really fit for the role;
> 
> Perhaps not but currently it is, what would an alternative look like?

I didn't make one alternative yet as I was trying to minimize changes in current code so I tried hard to not change the structures and not have to adapt every board driver. But the idea I have in mind is something like this:

```
typedef struct st_segment {
    uint_fast8_t id;                //!< Id may be used by driver to track changes
    struct st_segment *next;        //!< Pointer to next element in cirular list of segments
    st_block_t *exec_block;         //!< Pointer to the block data for the segment
    uint32_t cycles_per_tick;       //!< Step distance traveled per ISR tick, aka step rate.
    float current_rate;
    float target_position;          //!< Target position of segment relative to block start, used by spindle sync code
    uint_fast16_t n_step;           //!< Number of step events to be executed for this segment
#ifdef SPINDLE_PWM_DIRECT
    uint_fast16_t spindle_pwm;      //!< Spindle PWM to be set at the start of segment execution
#else
    float spindle_rpm;              //!< Spindle RPM to be set at the start of the segment execution
#endif
    bool update_rpm;                //!< True if set spindle speed at the start of the segment execution
    bool spindle_sync;              //!< True if block is spindle synchronized
    bool cruising;                  //!< True when in cruising part of profile, only set for spindle synced moves
    uint_fast8_t amass_level;       //!< Indicates AMASS level for the ISR to execute this segment

    // was in stepper_t
    uint32_t counter_x,             //!< Counter variable for the Bresenham line tracer, X-axis
             counter_y,             //!< Counter variable for the Bresenham line tracer, Y-axis
             counter_z              //!< Counter variable for the Bresenham line tracer, Z-axis
    #ifdef A_AXIS
           , counter_a              //!< Counter variable for the Bresenham line tracer, A-axis
    #endif
    #ifdef B_AXIS
           , counter_b              //!< Counter variable for the Bresenham line tracer, B-axis
    #endif
    #ifdef C_AXIS
           , counter_c              //!< Counter variable for the Bresenham line tracer, C-axis
    #endif
    #ifdef U_AXIS
           , counter_u              //!< Counter variable for the Bresenham line tracer, U-axis
    #endif
    #ifdef V_AXIS
           , counter_v              //!< Counter variable for the Bresenham line tracer, V-axis
    #endif
    ;
    uint32_t steps[N_AXIS];         //!< Number of step pulse event events per axis step pulse generated.
    uint_fast16_t steps_countdown;  //!< Steps remaining in line segment motion.
} segment_t;

//! Stepper ISR data struct. Run data.
typedef struct stepper {
    segment_t *exec_segment;        //!< Pointer to the segment beeing executed.
    //st_block_t *exec_block;       don't need this, as it is the same of exec_segment->exec_block
    //uint32_t step_event_count;     don't need this, as it is the same of exec_segment->n_step
    axes_signals_t *step_outbits;   //!< The stepping signals to be output. "1byte*steps_count"-sized array malloc'ed/free'd on new segment/segment done.
    uint32_t steps_write_index;     //!< Number of step pulse events ready for exec.
    uint32_t steps_read_index;      //!< Number of step pulse events already exec'ed.

    bool new_block;                 //!< Set to true when a new block is started, might be referenced by driver code for advanced functionality.
    bool new_segment;               //!< Set to true when a new segment is started, might be referenced by driver code for advanced functionality.
    uint_fast8_t amass_level;       //!< AMASS level for this segment.
//    uint_fast16_t spindle_pwm;
#ifdef ENABLE_BACKLASH_COMPENSATION
    bool backlash_motion;
#endif
    bool dir_change;                //!< Set to true on direction changes, might be referenced by driver for advanced functionality.
    axes_signals_t dir_outbits;     //!< The direction signals to be output. The direction signals may be output only when \ref stepper.dir_change is true to reduce overhead.
} stepper_t;
```

Basically Bresenham and AMASS data is fully contained in segment_t because some vars have been moved from stepper_t to segment_t, as well as some redundant vars are removed (ex: exec_block, step_event_count) because they already exist in segment_t and the stepper_t has the segment_t pointer.

> > a left over from grbl conversion to HAL'lized grbl?
> 
> Yes, and I have added fields to it and exposed it (making it accessible via the hal.stepper.pulse_start parameter).
> 
> > Any chance to change it?
> 
> Yes, but you will have to move code from the HAL to the core to make spindle synced motion work? And new entries has to be added to the HAL struct for the plugins now hooking into hal.stepper.pulse_start? It will be easier to run the relevant HAL/plugin code on the same core (as stepper.c)?

Awesome. I was hoping to have the chance of that.

The HAL must provide access to hardware and being core logic agnostic. So the spindle code now existing in HAL works but isn't in the right place. It isn't a multicore issue, it's an hardware abstraction error. But yes, the mod might become mandatory once we go multicore/multimcu.

Once we have the HAL running on core0 and CORE running on core1, the hal must not access core data (and core must not access hal data) in order to avoid both concurrency issues and the required access control set in place to avoid concurrencies issues (ie: semaphores, mutexes, critical sections; they are pretty efficient but still... add little cpu cost and delays if data must be accessed cross-core). To make an example: the serial input buffer must be written by HAL and read by CORE, so the buffer indexes must be protected by a mutex or alike, every other HAL data instead, must be HAL-only. The same applies for gpios and any other peripheral.

In the case of the spindle, I didn't study the code, but looks like changing the PWM freq/period in order to change the spindle rotation speed. So the HAL must provide set_pwm(freq, period) and everything else must go in grbl core or plugin. Then register set_pwm(freq, period) as an hal function pointer and let the grbl core (or plugin) manage the spindle using the function pointer. It also simplify the board drivers all, as a new board driver don't have to know how to spindle works... just provide the pwm function.

**Reply by terjeio** [reply]
> Basically Bresenham and AMASS data is fully contained in segment_t because some vars have been moved from stepper_t to segment_t

Don't do that, it is better to split them out to a new struct - possibly a static instatiation in stepper_driver_interrupt_handler(). The segment_t struct has many instances (the segment buffer) and either you have to dereference the variables via a pointer or make a local copy of the one currently executing. Pointer dereferencing in ISRs should be kept low.

> ... the hal must not access core data (and core must not access hal data) ...

I do not get this, I guess you mean access via pointers? Data has to be somehow moved across the cores or nothing can work.

Which parts of the code should run on which core should be determined. IMO the code for running the motors is what needs to be moved, at least as the initial exercise. So at least planner.c and stepper.c and related driver code as this is the real-time stuff that is critical and is where the challenge lies. 

>  just provide the pwm function.

It is not as simple as that, the grblHAL core needs to get the real-time angular position from somewhere - this means hardware access.

**Reply by mfp20** [reply]
> > Basically Bresenham and AMASS data is fully contained in segment_t because some vars have been moved from stepper_t to segment_t
> 
> Don't do that, it is better to split them out to a new struct - possibly a static instatiation in stepper_driver_interrupt_handler(). The segment_t struct has many instances (the segment buffer) and either you have to dereference the variables via a pointer or make a local copy of the one currently executing. Pointer dereferencing in ISRs should be kept low.

Yeah, you are right; the st_block_t and segment_t data might be discarded at computing time, ie: before execution, so everything needed at exec time must be copied in the final stepper_t.

> > ... the hal must not access core data (and core must not access hal data) ...
> 
> I do not get this, I guess you mean access via pointers? Data has to be somehow moved across the cores or nothing can work.

My bad. Ofc data must be accessed, but in the case of multiple cores accessing the same data, there must be sync tools (semaphores, mutexes, critical sections to disable IRQs, and so on) in place to avoid data corruption in the case of non-atomic accesses. And in turn, those sync tools should be as few as possible, because they introduce delays. This have to be done for every non-atomic (ie: non single instruction) operation on data. Because of this, there's the need of strict abstraction between core and hal, so that only the strictly needed data is used on both cores and often access happens by function (ie: not directly using the variables).

> Which parts of the code should run on which core should be determined. IMO the code for running the motors is what needs to be moved, at least as the initial exercise. So at least planner.c and stepper.c and related driver code as this is the real-time stuff that is critical and is where the challenge lies.

That part is actually pretty easy as it is a pipeline: gcode in, steps out. The only critical change is in the interrupt handler, but I did it already in the offloaded branch. I've just dropped that branch and restarted from your upstream code because in that branch I was using an array of stepper_t as buffer, but it doesn't really fit the role. Once I got the chance to mod stepper_t, I can buffer the step_outbits only, instead of buffering the whole stepper_t (multiple bytes) for every single step (8 bits); and avoid memcpy'ing the whole stepper_t for every step. So, by modding stepper_t, I spare both memory and cpu.

> > just provide the pwm function.
> 
> It is not as simple as that, the grblHAL core needs to get the real-time angular position from somewhere - this means hardware access.

So, add an HAL function to be called by core in order to get the angular position and ... any other needed method! In any case the "hardware" in "HAL" is the computing hardware (ie: pwm engine), not the application hardware (ie: spindle).

**Reply by mfp20** [reply]
> > Basically Bresenham and AMASS data is fully contained in segment_t because some vars have been moved from stepper_t to segment_t
> 
> Don't do that, it is better to split them out to a new struct

This is the best of the different approaches I tried in these days and still... I don't like it.

```
//
typedef union {
    uint8_t mask;
    uint8_t value;
    struct {
        uint8_t x :1,
                y :1,
                z :1,
                a :1,
                b :1,
                c :1,
                u :1,
                v :1;
    };
} st_exec_pos_change_t;

//
typedef struct st_exec_step_s
{
    bool dir_change;
    axes_signals_t dir_outbits;
    axes_signals_t step_outbits;
    st_exec_pos_change_t pos_outbits;
} st_exec_step_t;

//
typedef struct st_exec_segment_s
{
    uint_fast8_t id;
    bool start;

    // segment data used by interrupt handler
    uint32_t cycles_per_tick;
    uint_fast16_t n_step;
    bool update_rpm;
    #ifdef SPINDLE_PWM_DIRECT
    uint_fast16_t spindle_pwm;
    #else
    float spindle_rpm;
    #endif

    // segment data used by driver
    float target_position;
    bool spindle_sync;
    bool cruising;

    // steps
    st_exec_step_t *step;
} st_exec_segment_t;

//
typedef struct st_exec_block_s
{
    uint_fast8_t id;
    bool start;

    // block data used by interrupt handler
    gc_override_flags_t overrides;
    char *message;
    output_command_t *output_commands;
    bool dynamic_rpm;
    bool backlash_motion;

    // block data used by driver
    float steps_per_mm;
    float programmed_rate;

    // segments
    st_exec_segment_t *segment;
} st_exec_block_t;

```

There's simply too much stuff going on in both the interrupt handler and the board driver.

I can probably rework again the interrupt handler to reduce the amount of variables needed by the interrupt handler.
The board driver is an HAL, it should be happy with data in st_exec_step_t above; that data only. The spindle stuff counts many variables and there's no point to have those in the stepping interrupt handler; the spindle control can be way less precise in time than the precision needed by the stepping.

Probably I'm fighting with an overall architecture made for 8bit AVRs, ie: malloc/free and large arrays are impossible because of the hw constraints. So the motion pipeline has been stretched so that every single step can be computed in time using a fixed static memory pool, instead of have the algorithm more dynamic. Once we have 32bit wide ints and higher cpu freqs, we can easily manage the whole thing in a different way.

As an example the stepping, the messages and the output_commands can be coupled with timestamp and being triggered by an IRQ with precise timing, because having more IRQs on a faster mcu isn't an issue.

I'm thinking to completely rewrite the motion pipeline for offloaded operations, instead of reusing the same code for both monolithic and offloaded operation.

**Reply by terjeio** [reply]
I as I said before I believe planner.c, stepper.c and the related interrupt rutines should be all be offloaded to the second core - having stepper data transmitted across cores is IMO only going to slow things down. A clean and lean API between the remaining grblHAL core is what is needed?

> Once we have 32bit wide ints and higher cpu freqs, we can easily manage the whole thing in a different way.

grblHAL is not usable on an 8-bit architecture... There is one 16-bit driver (I would be surprised if anyone uses it), the rest is 32-bit. Still, some of the supported processors are memory constrained, both RAM and flash.

> using a fixed static memory pool

I plan to malloc at least the planner buffer, perhaps the segment buffer as well - this to make the size run-time configurable. For that linker scripts has to be modified and all the drivers retested... Both are now implemented as linked lists so the code change will be minimal.
I would not attempt introduce high frequency to malloc/free in the code - risking heap fragmentation...

> I'm thinking to completely rewrite the motion pipeline for offloaded operations, instead of reusing the same code for both monolithic and offloaded operation.

That is what I wanted/want to do eventually - by having an API that allows planner.c, stepper.c and the related interrupt rutines to have different implementations. As a first step planner.c, stepper.c and the related interrupt rutines can be reused, later on these could be moved to a FPGA, S-curve acceleration implemented and what not. Mach3 has what I belive is such an API roughly at the same place in the code - I briefly looked into it a few years ago. LinuxCNC also has one as there is at least one FPGA implementation - but I have not kooked at that. The Mach3 (and Mach4?) can transfer the needed data via USB or Ethernet (RPC?), dont't know what LinuxCNC does. There is a dev kit for the Mach3 implementation, IIRC with examples. LinuxCNC is open source.

**Reply by mfp20** [reply]
> I as I said before I believe planner.c, stepper.c and the related interrupt rutines should be all be offloaded to the second core - having stepper data transmitted across cores is IMO only going to slow things down. A clean and lean API between the remaining grblHAL core is what is needed?

The clean and lean API is already in place; it's fairly good, the only exception I see is the spindle code being in the board driver instead of the core.

You are referring to the [use case 2](https://raw.githubusercontent.com/mfp20/grblhk-kernel/offloaded/docs/use_cases.png): that's fairly simple as the irq triggers on the second core (or fpga), ie: the core where all the motion pipeline is computed. Once the small text gcode is handled to the second core, there's no computed motion data passing back to the first core; hal and sys calls only. Once all the calls to hal and sys are atomic (and they mostly are, didn't check them all yet), the job is over. 

Other use cases instead - ex: offloading to host - introduce major challenges.

> > Once we have 32bit wide ints and higher cpu freqs, we can easily manage the whole thing in a different way.
> 
> the rest is 32-bit. Still, some of the supported processors are memory constrained, both RAM and flash.

Sure they are, but having 4 bytes wide ints mean you can move 4 steps (8 motor each step) per cycle so the "tx protocol" is much faster. Or 4 chars in other circumstances (ex: 32bit DMA copy). Then there's the frequency: you have much more cycles per second, so you can use fat ISRs and still don't clog the main procedure or having jitter. Having this luxury opens the way to complex code architectures not available on 8bit AVRs... I've been trying for 1 year on those... no way...

In other words we are using a moped engine on a Ferrari frame.

> > using a fixed static memory pool
> 
> I plan to malloc at least the planner buffer, perhaps the segment buffer as well - this to make the size run-time configurable. For that linker scripts has to be modified and all the drivers retested... Both are now implemented as linked lists so the code change will be minimal.
> I would not attempt introduce high frequency to malloc/free in the code - risking heap fragmentation...

There's no real need of high freq malloc/free. You malloc once on boot a memory pool for your custom allocator and then manage the pool yourself. The allocator itself has some cpu overhead, but on AVRs it was impossible to think about it; on those faster mcus instead, you can have a simple linear allocator without missing the performance point... and a linear allocator doesn't fragment.

> > I'm thinking to completely rewrite the motion pipeline for offloaded operations, instead of reusing the same code for both monolithic and offloaded operation.
> 
> That is what I wanted/want to do eventually - by having an API that allows planner.c, stepper.c and the related interrupt rutines to have different implementations. As a first step planner.c, stepper.c and the related interrupt rutines can be reused, later on these could be moved to a FPGA, S-curve acceleration implemented and what not. 

As I wrote, use case 1 and 2 can be implemented using the grbl pipeline as is. Use cases 3-5 not really, because the data isn't parted in stages. The board driver (ie: last stage in the pipeline) uses data from the planner (ie: stage1)! There's no clean separation. Moreover this data is even copied in some cases; so there's cpu and ram waste. Again: that was meaningful on AVRs... 

And still I can't figure out why the board driver needs more than just 2 bits (dir+step) per motor per step.

> Mach3 has what I belive is such an API roughly at the same place in the code - I briefly looked into it a few years ago. LinuxCNC also has one as there is at least one FPGA implementation - but I have not kooked at that. The Mach3 (and Mach4?) can transfer the needed data via USB or Ethernet (RPC?), dont't know what LinuxCNC does. There is a dev kit for the Mach3 implementation, IIRC with examples. LinuxCNC is open source.

The code to be run on the capable host isn't really a problem. Once you have an RPi4 or an Intel/AMD cpu, you can use any kind of {cpu & fpu & ram}-intensive algos. I have a python3 re-write of Klipper's code (original one is python2); it has a motion planner for kartesian, corexy... can't remember exactly; I wrote it 2 years ago and then got stuck at the firmware for the mcu. Eventually, a plugin for LinuxCNC can be written in order to teach it how to send grbl-compatible blocks and segments...
Host's code isn't a problem at the moment.

**Reply by mfp20** [reply]
> I as I said before I believe planner.c, stepper.c and the related interrupt rutines should be all be offloaded to the second core

> Other use cases instead - ex: offloading to host - introduce major challenges.

Problem solved.

```
void st_prep_buffer()
{
    // Block step prep buffer, while in a suspend state and there is no suspend motion to execute.
    if (sys.step_control.end_motion)
        return;

    if (BOARD_OFFLOAD_TO_HOST) {

        // TODO pull blocks from host

        // Check if we need to fill the buffer.
        while (segment_buffer_tail != segment_next_head) {
            
            // TODO pull segments from host

            // Segment complete! Increment segment pointers, so stepper ISR can immediately execute it.
            segment_buffer_head = segment_next_head;
            segment_next_head = segment_next_head->next;
        }
    }
    else
    {
         [...existing code...]
    }
}
```

That's all needed in order to offload to external host. The external host must serialize outgoing st_block_t+segment_t, and mcu deserialize incoming ones. As st_prep_buffer() gets called every time the realtime system have spare time to spend, the buffers will be always full as long as the host is always ready to send. And the host is, as it has more cpu power.

The input part is trivial: a similar approach must be in gc_execute_block(): in the case of offloading to host, it does nothing (ie: doesn't run the planner code, as the blocks and segments are pulled directly in the existing buffers).

```
status_code_t gc_execute_block(char *block, char *message)
{
    if (BOARD_OFFLOAD_TO_HOST) {
        // do nothing
    }
    else if (BOARD_OFFLOAD_TO_CORE)
    {
        // TODO push to core
    }
    else
    {
         [...existing code...]
    }
}
```

There might be some 'hal' and 'sys' calls to add but that's the simplest way to implement offloading to host without introducing major changes to data structures. The only thing to test is ... wheter the interface in use (ex: UART, USB) is fast enough to transmit blocks and segments in time.

Gimmie some time and I'll complete the code both for offloading to second core (or fpga), and external host.

**Reply by terjeio** [reply]
> Problem solved.

Nearly? You offloaded far too much in your original approach, this is better. But I think planner.c should be offloaded too, `plan_buffer_line()` is where the bulk of the traffic across the cores should be? This as `st_update_plan_block_parameters()` will be called by the planner at a high frequency depending on the gcode.

IMO do not use if...then...else to determine where the data should be sent, use function pointers (or weak function declarations). Function pointers are the most flexible as they can be changed at run time, and if adding those needed to a new struct it will be clear where the crossover traffic takes place.

**Reply by mfp20** [reply]
> > Problem solved.
> 
> Nearly? You offloaded far too much in your original approach, this is better. 

Well, I was trying to offload ALL the math. This one is a compromise instead, because in the interrupt handler there's some math going on as well and it isn't offloaded. Those are just a bunch of sums and shifts, not much, but still... a compromise.
To get ALL the math offloaded there's the need to change data structures and ... we already talked about that.
So at the end of the day, I like this solution A LOT.

> But I think planner.c should be offloaded too, `plan_buffer_line()` is where the bulk of the traffic across the cores should be? This as `st_update_plan_block_parameters()` will be called by the planner at a high frequency depending on the gcode.

I don't get this part: in the case of BOARD_OFFLOAD_TO_HOST no gcodes should be sent to console, and in the case some gcode arrives there, gc_execute_block() simply does nothing except writing a warning in log/console. There are no calls to motion control, so none to planner as well.

> IMO do not use if...then...else to determine where the data should be sent, use function pointers (or weak function declarations). Function pointers are the most flexible as they can be changed at run time, and if adding those needed to a new struct it will be clear where the crossover traffic takes place.

`Currently changes are implemented using some defines in config.h, so the offloading is set at compile time to produce an experimental build. But the final version might be able to set offloading at runtime (ex: on boot).`
([README.md](https://github.com/mfp20/grblhk-kernel))
